integrations/clickup/client.py

```diff
@@ -0,0 +1,244 @@
+        yield (await self._send_api_request(f"{self.api_url}/team")).get("teams", [])
+
+    @cache_iterator_result()
+    async def _get_spaces_in_team(
+        self, team_id: str
+    ) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all spaces in a workspace."""
+        yield (
+            await self._send_api_request(
+                f"{self.api_url}/team/{team_id}/space",
+                {"is_archived": self.is_archived},
+            )
+        ).get("spaces", [])
+
+    @cache_iterator_result()
+    async def _get_folders_in_space(
+        self, team_id: str
+    ) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all folders in a space."""
+        async for spaces in self._get_spaces_in_team(team_id):
+            for space in spaces:
+                yield (
+                    await self._send_api_request(
+                        f"{self.api_url}/space/{space.get('id')}/folder",
+                        {"is_archived": self.is_archived},
+                    )
+                ).get("folders")
+
+    async def get_all_spaces(self) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all spaces across all teams."""
+        async for teams in self.get_clickup_teams():
+            for team in teams:
+                async for spaces in self._get_spaces_in_team(team.get("id")):
+                    yield [{**space, TEAM_OBJECT: team} for space in spaces]
+
+    async def get_folder_projects(self) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all projects with a folder parent."""
+        async for teams in self.get_clickup_teams():
+            for team in teams:
+                async for folders in self._get_folders_in_space(team.get("id")):
+                    for folder in folders:
+                        projects = (
+                            await self._send_api_request(
+                                f"{self.api_url}/folder/{folder.get('id')}/list",
+                                {"is_archived": self.is_archived},
+                            )
+                        ).get("lists")
+                        yield [{**project, TEAM_OBJECT: team} for project in projects]
+
+    async def get_folderless_projects(
+        self,
+    ) -> AsyncGenerator[List[dict[str, Any]], None]:
+        """Get all projects without a folder parent."""
+        async for teams in self.get_clickup_teams():
+            for team in teams:
+                async for spaces in self._get_spaces_in_team(team.get("id")):
+                    for space in spaces:
+                        projects = (
+                            await self._send_api_request(
+                                f"{self.api_url}/space/{space.get('id')}/list",
+                                {"is_archived": self.is_archived},
+                            )
+                        ).get("lists")
+                        yield [{**project, TEAM_OBJECT: team} for project in projects]
```

Comment:
This contains repeated code going through the teams

Reply:
It has been refactored